### ISSUE TITLE
Working with data of different lengths using automasking

### ISSUE THREAD

**Issue by noashin** [google/jax#5000]
I am working with multiple time series of different lengths, that should all undergo the same operations. 
Following the [tweet](https://twitter.com/marcvanzee/status/1305516283941867521) from @marcvanzee I attempted to use mask and vmap for these manipulations. Unfortunately due to the lack of documentation, I am not sure I can fully decipher Marc's example. 
The simplest operation I am attempting is concatenating two padded arrays. Is it possible to parallelize it using mask and vmap?
```python
import jax.numpy as np

# each row is a data_set of different length. Zero padding is used to put it all in one array.
data_sets_a = np.array([[1,2,0,0], [4,5,6,7]])
data_lengths_a = np.array([2, 4])

data_sets_b = np.array([[1,0], [1,2]])
data_lengths_b = np.array([1, 2])

num_data_sets = data_sets_a.shape[0]

# implementation using list comprehension
data_sets_concatenated_list = [np.concatenate([data_sets_a[i, :data_lengths_a[i]], data_sets_b[i, :data_lengths_b[i]]]) \
                               for i in range(num_data_sets)]
data_sets_concatenated = np.array(list(itertools.zip_longest(*data_sets_concatenated_list, fillvalue=0))).T
```

Is it possible to define something like:
```python
def concatenate(array_1, array_2):
    return np.concatenate([array_1, array_2])

input_mask_1 = data_lengths_a
input_mask_2 = data_lengths_b
output_mask = data_lengths_a + data_sets_b
```
and use vmap and mask to get the same result as in the code above?

**Comment by JossWhittle** [comment]
Hi noashin, I'm looking at similar stuff and I found this comment with an example of the masking feature https://github.com/google/jax/issues/2521#issuecomment-604759386

It's from march 2020 so I don't know if they've changed anything about the API since then. I'm trying to get it working myself too so please say if you manage to get it working :)

**Comment by noashin** [comment]
So far no luck. For the more basic operations I need I am using padded arrays, but for the more complicated ones, I am currently looping.

**Comment by myagues** [comment]
Not an *exact* answer, but looking at the masking tests and that comment:
```python
@functools.partial(jax.mask, in_shapes=["n", "m"], out_shape="_")
def mask_concat(x, y):
    return jnp.pad(jnp.concatenate([x, y]), (0, 1))

foo = jax.vmap(mask_concat)([data_sets_a, data_sets_b], dict(n=data_lengths_a, m=data_lengths_b))[:, :-1]

> [[1 2 1 0 0 0]
   [4 5 6 7 1 2]]
```
Maybe there is something more involved and pretty that can get the job without that final slicing, but it should work.

The explanation is that `mask_concat` will receive two 1d arrays (because of `vmap`) and the sizes to take into account will be defined by the values in `n` and `m`, respectively. The resulting 1d array will have an undefined length (`n + m + 1`, but we can skip setting it explicitly).

**Comment by noashin** [comment]
@myagues Thank you so much! This is definitely what I need.
I tried to apply it to an example where the function to be masked and vmapped operates on a vector and a matrix, and I can't get it to work.
```python
import jax.numpy as jnp
from jax import mask, vmap

def my_func(vec, mat):
    return jnp.dot(mat, vec)

mat = jnp.ones((2, 3, 3))
vec = jnp.array([[1., 2., 3.], [1., 2., 0.]])

inds = jnp.array([3, 2])
inds_mat = jnp.array([[2, 2], [3, 3]])

masked_func = mask(my_func, in_shapes=["n", "m"], out_shape="_")
vmapped_masked_func = vmap(masked_func, (0, 0), 0)
res = vmapped_masked_func([vec, mat], dict(n=inds, m=inds_mat))
```

The error is `FilteredStackTrace: AssertionError: length mismatch: [1, 2]`. 
I think the problem is in the definition of the masking indices of the matrix `inds_mat`, but I don't have an idea how to solve it.

**Comment by myagues** [comment]
I am not sure if it is possible to index with "multi index" as in `inds_mat`. I have not found anything of the sorts, but there is an example in the tests with `lax.dot`:
https://github.com/google/jax/blob/4d78d60a24f2cfd2b16a3cf8a2992b693ff1de05/tests/masking_test.py#L432-L437

The error you are getting is a length mismatch between `mat` and the `in_shapes` value you are declaring: `"m"` should be `"(m, _)"`. Essentially, you are giving a matrix, but `jax.mask` is expecting a vector.

Since we are using `jnp.dot`, that second dimension needs to be a fixed value resulting in: `in_shapes=["n", "(m, n)"]`. However, I guess `m` takes into account **only** the leading value of `inds_mat`? Also, the resulting matrix will give "padded" values for shorter sequences:
```python
masked_func = jax.mask(my_func, in_shapes=["n", "(m, n)"], out_shape="_")
jax.vmap(masked_func)([vec, mat], dict(n=inds, m=inds_mat))
> DeviceArray([[6., 6., 6.],
               [3., 3., 3.]], dtype=float32)
```

**Comment by mattjj** [comment]
It's really great to see folks helping each other here! Actually it's impressive what you've figured out just by reading the code :) I'm sorry that I haven't been able to provide any useful help so far.

We're redesigning `mask`. That redesign, together with the fact that we've got our hands full with other tasks and support requests, means that we've made improving and documenting the current `mask` API a low priority (to the point where, in practice, we never get to responding in threads like this one, as much as we'd like to help).

Until we get our designs sorted out, I wouldn't recommend using `mask` unless you're just having fun experimenting. In addition to being undocumented, its implementation is incomplete and the error messages it gives are terrible, as you've already seen. And I don't expect them to get better in the near future, pending the redesign.

If you are just having fun experimenting, as for the question in the OP:

> I am working with multiple time series of different lengths, that should all undergo the same operations.

Check out the "ragged batching" section of [this gist](https://gist.github.com/mattjj/95438e43c3bf0eef7c3bf0235f314336). That shows a way to use `vmap` + `mask` together to operate on multiple time series of different lengths (using a `scan`). Instead of being concatenated/packed, the sequences are padded out to the same length and then operations on them are masked appropriately. That is, in terms of how many total FLOPs are expended, all sequences effectively have the maximum length. This approach can still be profitable on a parallel machine like a hardware accelerator, but I wanted to underscore that it's different from the concatenate/pack approach that I think the discussion in this thread is focused on (which is the approach you might see for RNNs on GPUs, for example).

To summarize, I don't recommend using `mask`, but we are working on this area and hope to provide some new solutions in the coming months. In the meantime, it's really helpful to have users raise issues like this, since it helps guide our work! (That's actually how I just rediscovered this thread: I was mining our issues page for examples that our new design must be able to handle.)

**Comment by krzysztofrusek** [comment]
@mattjj Thank you for all the detailed comments here.

I am starting my adventure with JAX and other great libraries built around it.
In particular, I want to test Jraph library for GNNs, however, the example of masking they have throws an exception about broken masking in JAX.
Since the mask is not recommended right now I wanted to implement efficient masking by hand.

TLDR I need to apply  NN to padded batches (I have a boolean mask).
Right now I am using `jnp.where` to select masked parts and zeros for padding to save computations.

I am curious if this will be properly jitted to avoid the materialization of extra tensors of zeros?
Or maybe there is a better way of doing handmade masking?

**Comment by NightMachinary** [comment]
I have the opposite problem; I have some fixed length arrays, and I want to drop a postfix of them (i.e., ignoring batching, `a.at[x:].set(0)`)(to make the network work with variable length inputs, and reduce overfitting). What's the efficient way to do this?

**Comment by mattjj** [comment]
We deleted `mask`, and the discussion here is almost 3 years old, so this issue is stale. If folks have questions about how to implement padded-and-masked computations, don't hesitate to open new issues! But I'm closing this one.
